### PR TITLE
WIP: Update Example App to angular v7, angular-redux v9

### PR DESCRIPTION
**Pull request by wtho**
This PR tackles and maybe even closes several issues (#54, #47, #43, #40, angular-redux/example-app#49 and mostly #8).

I did not change any of the logic.

I am aware Angular v7 does fulfill the peer dependency of `angular-redux/store`, but I know it works and I wanted to push the example app as far as possible, knowing that it might catch some dust again.
If this is an issue, the dependency can be downgraded to Angular v6.

An example of Lazy-Loading is not implemented in this PR.

**Updates**
  * Dependency angular-redux v9 and angular v7
  * Use Angular-cli v7 and angular.json instead of angular-cli.json
  The Config is not completely configured with tslint and integrated tests, I just took the generic config to make the app run with the current angular-cli
  * Refactor code to use rxjs v6
  * Use Jest as Unit-Test Library with jest-preset-angular
  This is a personal choice, but I saw you use jest in the root of this repository as well. For jasmine there are plenty of examples, whereas for jest and angular it keeps being shaky.
  * Improve type safety complying to Strict Class Initialization and removing implicit any
  * Avoiding Circular Dependency by moving action strings to ticket-actions.ts
  * Update @angular/http to @angular/common/http (HttpClient)
  * Change mocky.io links to https, [taken from angular-redux/example-app PR](https://github.com/angular-redux/example-app/pull/52), kudos to @antonshevyrin

### CODE REVIEW

**Comment by smithad15** [comment]
This is great. Thank you very much for this. I will take a moment to review the code soon, but right now I'm a bit confused as to why the CircleCI hooks aren't firing. Is your branch up to date with the latest from `master`?

**Comment by smithad15** [comment]
Looks like its an issue with how I setup npm auth tokens (https://circleci.com/gh/wtho/platform/2?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link). I'll look into it

File: packages/example-app/src/app/animals/animal/component.ts [hunk 1]
```diff
@@ -22,20 +24,29 @@ export const toSubTotal = (obs$: Observable<Animal>): Observable<number> =>
...
+
+  /*
+  addTicket() {
+    this.actions.addTicket();
+  }
+
+  removeTicket() {
+    this.actions.removeTicket();
+  }
+  */
```
Comment (wtho) [inline packages/example-app/src/app/animals/animal/component.ts]: Dead code should be removed

**Comment by jamesbs** [comment]
> Looks like its an issue with how I setup npm auth tokens (https://circleci.com/gh/wtho/platform/2?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link). I'll look into it
Probably a good idea to add .npmrc to .gitignore

File: packages/example-app/package.json [hunk 2]
```diff
@@ -15,30 +18,35 @@
...
+    "@types/ramda": "^0.25.42",
     "@types/redux-logger": "^3.0.0",
-    "protractor": "~5.1.1",
-    "ts-node": "~3.0.2",
-    "typescript": "^2.4.1"
+    "jest": "^23.6.0",
+    "jest-preset-angular": "^6.0.1",
+    "protractor": "^5.4.1",
+    "ts-node": "^7.0.1",
+    "typescript": "^3.1.6"
```
Comment (jamesbs) [inline packages/example-app/package.json]: This is pulling in versions outside of the range required by @angular/compiler-cli. Best to just use the same range I think `">=3.1.1 <3.2"`

File: packages/example-app/src/app/animals/animal/component.ts [hunk 3]
```diff
@@ -22,20 +24,29 @@ export const toSubTotal = (obs$: Observable<Animal>): Observable<number> =>
   changeDetection: ChangeDetectionStrategy.OnPush,
 })
 export class AnimalComponent {
-  static readonly ADD_TICKET = 'ADD_TICKET';
-  static readonly REMOVE_TICKET = 'REMOVE_TICKET';
 
-  @Input() key: string;
-  @Input() animalType: string;
+  @Input() key!: string;
```
[the remark this answers is not in the document]
Comment (smithad15) [reply packages/example-app/src/app/animals/animal/component.ts]: I have traditionally taken the "just disable it" path before, but I admit that was probably mostly out of being lazy. Curious as to whether we need to comply with enabling that setting so as not to cause issues in consuming code bases that have that check enabled? Open to suggestions

[the remark this answers is not in the document]
Comment (wtho) [reply packages/example-app/src/app/animals/animal/component.ts]: So in the angular.io app ([example component](https://github.com/angular/angular/blob/master/aio/src/app/layout/mode-banner/mode-banner.component.ts)), in all examples on angular.io and in the downloadable example app [Tour of Heroes](https://angular.io/tutorial) the Angular team does disable it as well.

I would suggest to disable it as well. Will update with the next commit.

**Comment by wtho** [comment]
I do not know why the error with the `index.d.ts` in `@types/ramda` happened, but setting the version to 0.24 fixes it. Hope this is getting solved once `@types/ramda@0.26` gets released.

Now the test fails, but I do not know why exactly. Seems to be an issue with yarn:
* If I install the dependencies using `npm install` in `example-app` and run the tests, they pass
* If I install the dependencies of the example app in an isolated folder (not inside `angular-redux/platform`) with yarn and run the tests, they pass
* If I install with yarn inside `angular-redux/platform/packages/example-app` and run the tests, they fail. Somehow the installation depends on the `../../yarn.lock` in the root folder. I do like yarn for its speed, but if it does things differently in a non-transparent way, it can be quite annoying.
* If I run `yarn install` in the root repository, it installs most packages in the example app, but four are missing in `node_modules`: `core-js`, `flux-standard-action`, `ramda` and `zone.js` and the versions of many packages differ compared to the isolated example-app installation.

I do not know enough about how `yarn install` works with nested projects.

File: packages/example-app/src/app/elephants/page.spec.ts [hunk 4]
```diff
@@ -61,59 +59,51 @@ xdescribe('Elephant Page Container', () => {
...
 
-    elephantPage.animals$
-      .toArray()
+    elephantPage.animals$.pipe(toArray())
       .subscribe(
-        actualSequence => expect(actualSequence).toEqual(expectedSequence),
-        null,
-        done,
+        actualSequence => 
+          expect(actualSequence).toEqual(expectedSequence)
```
[the remark this answers is not in the document]
Comment (wtho) [reply packages/example-app/src/app/elephants/page.spec.ts]: They did not run. I added Promises to make sure the checks are run.

**Comment by jamesbs** [comment]
I would recommend only running scripts from the workspace folder rather than the individual package folders. This includes installs. I did some work in wtho/platform#2 to get all of the versions synced up. For any package that is required in different places, we need to make sure those versions are the same otherwise we'll run into issues with code executing in different package versions causing defects.

If you are having installation issues, executing a `clean` and reinstall may help.

**Comment by jamesbs** [comment]
Updated the base branch to merge into `upgrade-angular-7`. Considering how many changes will need to be involved I'd like to separate the changes into multiple commits before moving forward with a new release.

**Comment by wtho** [comment]
Yeah, I totally agree, there are many changes in different places happening here.

>  I'd like to separate the changes into multiple commits

Do you mean multiple PRs or multiple commits inside this PR (and no squashing to keep the history)?

I can clean the current changes in several commits up a bit if you want, reorganizing the changes by concerns.

**Comment by jamesbs** [comment]
I meant multiple PRs, while squashing and merging incoming commits to the upgrade branch.

**Pull request merged**